src/tql/query.py

```diff
@@ -0,0 +1,185 @@
+        else:
+            session_type = SessionType.DATA
+    else:
+        if args.session.lower() == "schema":
+            session_type = SessionType.SCHEMA
+        elif args.session.lower() == "data":
+            session_type = SessionType.DATA
+        else:
+            raise ArgumentError("Incorrect session type provided. Session type must be 'schema' or 'data'.")
+
+    if args.transaction is None:
+        if strict_transactions:
+            raise ArgumentError("Strict transaction types is enabled and no transaction type was provided. Use -t to specify transaction type.")
+        elif query_type in ("define", "undefine", "insert", "update", "delete"):
+            transaction_type = TransactionType.WRITE
+        else:
+            transaction_type = TransactionType.READ
+    else:
+        if args.transaction.lower() == "read":
+            transaction_type = TransactionType.READ
+        elif args.transaction.lower() == "write":
+            transaction_type = TransactionType.WRITE
+        else:
+            raise ArgumentError("Incorrect transaction type provided. Transaction type must be 'read' or 'write'.")
+
+    if args.inference is None:
+        options = typedb.client.TypeDBOptions().core().set_infer(global_inference)
+    else:
+        options = typedb.client.TypeDBOptions().core().set_infer(args.inference)
+
+    connection.set_session(session_type)
+
+    try:
+        with connection.session.transaction(transaction_type, options) as transaction:
+            if query_type == "match":
+                results = decode(transaction.query().match(query), ConceptMap)
+            elif query_type == "match_aggregate":
+                results = decode(transaction.query().match_aggregate(query).get(), Numeric)
+            elif query_type == "match_group":
+                results = decode(transaction.query().match_group(query), ConceptMapGroup)
+            elif query_type == "match_group_aggregate":
+                results = decode(transaction.query().match_group_aggregate(query), NumericGroup)
+            elif query_type == "define":
+                transaction.query().define(query)
+            elif query_type == "undefine":
+                transaction.query().undefine(query)
+            elif query_type == "insert":
+                transaction.query().insert(query)
+            elif query_type == "delete":
+                transaction.query().delete(query)
+            elif query_type == "update":
+                transaction.query().update(query)
+
+            if transaction_type == TransactionType.WRITE:
+                transaction.commit()
+                print('{} query success.'.format(query_type.title()))
+                return
+            else:
+                return results
+    except TypeDBClientException as exception:
+        if session_type == SessionType.SCHEMA:
+            print('Switching to data session due to failed schema query.')
+            connection.set_session(SessionType.DATA)
```

Comment:
This is necessary, as far as I'm aware, to solve a problem with sessions not being closed when exceptions are thrown. In order to arbitrarily persist sessions across cell blocks, they have to be opened without context managers. This means that there's no way to automatically close sessions with `__exit__`, so they have to be closed with `__del__`, which is correctly called either when the user closes a connection or the kernel is stopped, but not if an exception is thrown. This means that an exception has the potential to leave orphaned sessions open, which will leave orphaned locks if the session is a schema session, hence the change to data session on exception. I know this isn't ideal as it will still leave orphaned data sessions open, but it's the best I could think of. Auto-closing connections on exception is not really viable in my opinion as the user would have to then run cells out of order or restart the whole notebook (which could be very large) in order to modify and re-run the erroneous cell, which is not very Jupyteresque. Very open to suggestions on this one.